### DISCUSSION TITLE
[Debugger] Adding IntelliSense support for debug console

### DISCUSSION THREAD

**Discussion by praveennadarajah** [ballerina-platform/ballerina-lang#32285] [Ideas]
Currently, ballerina does not have IntelliSense (code-completion) support for debug console. Therefore users have to type the entire expression manually to evaluate it. Typing lengthy expressions in debug console without code completion would lead to a bad user experience.

How about adding Intellisense support for debug console?

We can **consider** the following approaches:

1. Writing a new completion engine from scratch in debug adapter (DA).
2. Reusing the language server (LS) utilities in the debug adapter and get the completion items.
3. Calling the language server from the debug adapter (with LS instance), get the response from LS, do the necessary conversions in DA and send the response back to VS Code.  In this approach, the debug adapter will act as the client and the language server will act as the server.

@hevayo @mohanvive @nadeeshaan @prabushi @NipunaRanasinghe @IMS94

**Comment by praveennadarajah** [comment]
Based on the internal discussion, it is decided to do an initial POC with **Approach 3**.

**Reply by praveennadarajah** [reply]
> Based on the internal discussion, it is decided to do an initial POC with **Approach 3**.

In **Approach 1**, there can be inconsistencies between LS and DA completions. Also having two completion engines in LS and DA is not appropriate.

In **Approach 2**, LS utilities are tightly coupled with the LS components, hence cannot be used directly.

Therefore **Approach 3** is selected. But this approach would be complex to handle asynchronous calls, and can be slow.

**Comment by praveennadarajah** [comment]
Please find the architecture design for the proposed **Approached 3** in [1].

[1] https://docs.google.com/document/d/1jGVvhyGLdA_KNdXnSSlcAkSYJIZpkMJr0boFLDrsC7A/edit?usp=sharing

**Reply by IMS94** [reply]
@praveennadarajah Design looks ok. One thing to note is that, before you send the **completion** request to LS (from DA), you have to make sure the document being edited is opened in LS and updated with the current content. That is, before sending the **completion** request, 
1. Upon the first interaction with expression editor, send a **textDocument/didOpen** event with the current content of the file. Current content means the actual file content + any lines added by DA with user's expression.
2. For each subsequent edit user make, send a **textDocument/didChange** event with the content of the file
3. Then, send the **textDocument/completion** request and return the response after necessary transformations

**Reply by praveennadarajah** [reply]
@IMS94 Thanks for your inputs. I will consider the above suggestions during the initial PoC.

**Comment by nadeeshaan** [comment]
The architecture diagram looks good. Also, I would like to suggest you the following points as well.
1. During the initialization of the language server, make sure to disable the unnecessary operations so the language server will not initialize and send requests and the notifications which cannot be handled by the client - DAP internal in our case
2. We need to figure out a way on how to capture the text changes added in the debug console. IINM, DAP does not have a text synchronisation event. (Please correct me if this statement is invalid)
